source/s3_util.c

```diff
@@ -307,3 +314,161 @@ void aws_s3_add_user_agent_header(struct aws_allocator *allocator, struct aws_ht
+int aws_s3_parse_content_length_response_header(
+    struct aws_allocator *allocator,
+    struct aws_http_headers *response_headers,
+    uint64_t *out_content_length) {
+    AWS_PRECONDITION(allocator);
+    AWS_PRECONDITION(response_headers);
+    AWS_PRECONDITION(out_content_length);
+
+    struct aws_byte_cursor content_length_header_value;
+
+    if (aws_http_headers_get(response_headers, g_content_length_header_name, &content_length_header_value)) {
+        aws_raise_error(AWS_ERROR_S3_MISSING_CONTENT_LENGTH_HEADER);
+        return AWS_OP_ERR;
+    }
+
+    struct aws_string *content_length_header_value_str =
+        aws_string_new_from_cursor(allocator, &content_length_header_value);
+    char *content_length_str_end = NULL;
+
+    *out_content_length = strtoull((const char *)content_length_header_value_str->bytes, &content_length_str_end, 10);
+    aws_string_destroy(content_length_header_value_str);
+
+    return AWS_OP_SUCCESS;
+}
+
+/* Calculate the number of parts based on object range and end. */
+uint32_t aws_s3_get_num_parts(size_t part_size, uint64_t object_range_start, uint64_t object_range_end) {
+    if ((object_range_start - object_range_end) == 0ULL) {
+        return 0;
+    }
+
+    uint32_t num_parts = 1;
+
+    uint64_t first_part_size = part_size;
+    uint64_t first_part_alignment_offset = object_range_start % part_size;
+
+    /* If the first part size isn't aligned on the assumed part boundary, make it smaller so that it is. */
+    if (first_part_alignment_offset > 0) {
+        first_part_size = part_size - first_part_alignment_offset;
+    }
+
+    uint64_t second_part_start = object_range_start + first_part_size;
+
+    /* If the range has room for a second part, calculate the additional amount of parts. */
+    if (second_part_start <= object_range_end) {
+        uint64_t aligned_range_remainder = object_range_end - second_part_start;
+        num_parts += (uint32_t)(aligned_range_remainder / (uint64_t)part_size);
+
+        if ((aligned_range_remainder % part_size) > 0) {
+            ++num_parts;
+        }
+    }
+
+    return num_parts;
+}
+
+/* Calculate the part range for a given part. Intended to be used in conjunction with aws_s3_get_num_parts. */
+void aws_s3_get_part_range(
+    uint64_t object_range_start,
+    uint64_t object_range_end,
+    size_t part_size,
+    uint32_t part_number,
+    uint64_t *out_part_range_start,
+    uint64_t *out_part_range_end) {
+    AWS_PRECONDITION(out_part_range_start);
+    AWS_PRECONDITION(out_part_range_end);
+
+    uint64_t part_size_uint64 = (uint64_t)part_size;
+    uint64_t first_part_size = part_size_uint64;
+    uint64_t first_part_alignment_offset = object_range_start % part_size_uint64;
+
+    AWS_ASSERT(part_number > 0);
+
+    const uint32_t part_index = part_number - 1;
+
+    /* Part index is assumed to be in a valid range. */
+    AWS_ASSERT(part_index < aws_s3_get_num_parts(part_size, object_range_start, object_range_end));
```

Comment:
Trivial: can you move those checks to the beginning of the function? Right after the `AWS_PRECONDITION(out_part_range_end);`